File: VindsidenKit/CDPlot.m
//
//  CDPlot.m
//  Vindsiden
//
//  Created by Ragnar Henriksen on 17.09.10.
//  Copyright (c) 2010 Shortcut AS. All rights reserved.
//

#import "CDPlot.h"
#import "CDStation.h"
#import "NSString+fixDateString.h"
#import "NSNumber+Between.h"

#import <VindsidenKit/VindsidenKit-Swift.h>

@implementation CDPlot

@dynamic plotTime;
@dynamic windAvg;
@dynamic windDir;
@dynamic windMax;
@dynamic windMin;
@dynamic tempWater;
@dynamic tempAir;

@dynamic station;


+ (CDPlot *) newOrExistingPlot:(NSDictionary *)dict forStation:(CDStation *)station inManagedObjectContext:(NSManagedObjectContext *)managedObjectContext
{
    CDPlot *existing = nil;
    NSString *dateString = [dict[@"plotTime"] fixDateString];
    NSDate *date = [[Datamanager sharedManager] dateFromString:dateString];
    NSFetchRequest *request = [[NSFetchRequest alloc] initWithEntityName:@"CDPlot"];

    request.predicate = [NSPredicate predicateWithFormat:@"station == %@ and plotTime == %@", station, date];
    request.fetchLimit = 1;

    NSArray *array = [managedObjectContext executeFetchRequest:request error:nil];

    if ( [array count] > 0 ) {
        existing = array[0];
    } else {
        existing = [[CDPlot alloc] initWithEntity:request.entity insertIntoManagedObjectContext:managedObjectContext];

        for (id key in dict ) {
            id v = dict[key];
            if ( [v class] == [NSNull class] ) {
                continue;
            } else if ( [key isEqualToString:@"plotTime"] ) {
                existing.plotTime = date;
                continue;
            } else if ( [key isEqualToString:@"windDir"]  ) {
                CGFloat value = [dict[key] floatValue];
                if ( value < 0 ) {
                    value = value + 360;
                }
                [existing setValue:@(value)
                            forKey:key];
                continue;
            } else if ( [key isEqualToString:@"stationID"] ) {
                continue;
            } else if ( [key isEqualToString:@"windMin"] ) {
                CGFloat value = [dict[key] floatValue];
                if ( value < 0 ) {
                    value = 0.0;
                }
                [existing setValue:@(value) forKey:key];
                continue;
            }
            [existing setValue:@([dict[key] floatValue]) forKey:key];
        }
    }

    return existing;
}


+ (void)updatePlots:(NSArray *)plots completion:(void (^)(void))completion
{
    if ( 0 == [plots count] ) {
        if ( completion ) {
            completion();
        }

        return;
    }

    NSManagedObjectContext *context = [[Datamanager sharedManager] managedObjectContext];
    NSManagedObjectContext *childContext = [[NSManagedObjectContext alloc] initWithConcurrencyType:NSPrivateQueueConcurrencyType];
    childContext.parentContext = context;
    childContext.mergePolicy = NSMergeByPropertyObjectTrumpMergePolicy;
    childContext.undoManager = nil;
    __block NSError *err = nil;

    [childContext performBlock:^{
        CDStation *thisStation = [CDStation existingStation:plots[0][@"stationID"] inManagedObjectContext:childContext];

        if ( thisStation == nil ) {
            [Logger DLOG:[NSString stringWithFormat:@"StationID %@ not found", plots[0][@"stationID"]] file:@"*" function:@(__PRETTY_FUNCTION__) line:__LINE__];

            if ( completion ) {
                completion();
            }

            return;
        }

        NSMutableSet *insertedPlots = [NSMutableSet set];

        for ( NSDictionary *dict in plots ) {
            CDPlot *managedObject = [CDPlot newOrExistingPlot:dict forStation:thisStation inManagedObjectContext:childContext];
            if ( [managedObject isInserted] ) {
                [insertedPlots addObject:managedObject];
            }
        }

        if ( insertedPlots.count ) {
            [thisStation willChangeValueForKey:@"plots"];
            [thisStation addPlots:insertedPlots];
            [thisStation didChangeValueForKey:@"plots"];
        }

        if ( [childContext hasChanges] && [childContext save:&err] == NO ) {
            [Logger DLOG:[NSString stringWithFormat:@"Save failed: %@", err.localizedDescription] file:@"*" function:@(__PRETTY_FUNCTION__) line:__LINE__];
            [Logger DLOG:[NSString stringWithFormat:@"Save failed: stationID: %@ - %@", plots[0][@"stationID"], thisStation] file:@"*" function:@(__PRETTY_FUNCTION__) line:__LINE__];
        }

        [childContext processPendingChanges];

        [context performBlock:^{
            if ( [context hasChanges] && [context save:&err] == NO ) {
                [Logger DLOG:[NSString stringWithFormat:@"Save failed: %@", err.localizedDescription] file:@"*" function:@(__PRETTY_FUNCTION__) line:__LINE__];
            }

            [context processPendingChanges];


            if ( completion ) {
                completion();
            }
        }];
    }];
}


- (NSString *)windDirectionString
{
    CGFloat direction = [self.windDir floatValue];

    if ( direction > 360.0 || direction < 0 ) {
        direction = 0.0;
    }
    NSNumber *dir = @(direction);


    if ( [dir isBetween:0.0 and:11.25] || [dir isBetween:348.75 and:360.001]) {
        return NSLocalizedString(@"DIRECTION_N", @"N");
    } else if ( [dir isBetween:11.25 and:33.35] ) {
        return NSLocalizedString(@"DIRECTION_NNE", @"NNE");
    } else if ( [dir isBetween:33.75 and:56.25] ) {
        return NSLocalizedString(@"DIRECTION_NE", @"NE");
    } else if ( [dir isBetween:56.25 and:78.75] ) {
        return NSLocalizedString(@"DIRECTION_ENE", @"ENE");
    } else if ( [dir isBetween:78.75 and:101.25] ) {
        return NSLocalizedString(@"DIRECTION_E", @"E");
    } else if ( [dir isBetween:101.25 and:123.75] ) {
        return NSLocalizedString(@"DIRECTION_ESE", @"ESE");
    } else if ( [dir isBetween:123.75 and:146.25] ) {
        return NSLocalizedString(@"DIRECTION_SE", @"SE");
    } else if ( [dir isBetween:146.25 and:168.75] ) {
        return NSLocalizedString(@"DIRECTION_SSE", @"SSE");
    } else if ( [dir isBetween:168.75 and:191.25] ) {
        return NSLocalizedString(@"DIRECTION_S", @"S");
    } else if ( [dir isBetween:191.25 and:213.75] ) {
        return NSLocalizedString(@"DIRECTION_SSW", @"SSW");
    } else if ( [dir isBetween:213.75 and:236.25] ) {
        return NSLocalizedString(@"DIRECTION_SW", @"SW");
    } else if ( [dir isBetween:236.25 and:258.75] ) {
        return NSLocalizedString(@"DIRECTION_WSW", @"WSW");
    } else if ( [dir isBetween:258.75 and:281.25] ) {
        return NSLocalizedString(@"DIRECTION_W", @"W");
    } else if ( [dir isBetween:281.25 and:303.75] ) {
        return NSLocalizedString(@"DIRECTION_WNW", @"WNW");
    } else if ( [dir isBetween:303.75 and:326.25] ) {
        return NSLocalizedString(@"DIRECTION_NW", @"NW");
    } else if ( [dir isBetween:326.25 and:348.75] ) {
        return NSLocalizedString(@"DIRECTION_NNW", @"NNW");
    } else {
        return NSLocalizedString(@"DIRECTION_UKN", @"UKN");
    }
}


@end

File: VindsidenKit/CDPlot.swift
//
//  CDPlot.swift
//  VindsidenApp
//
//  Created by Ragnar Henriksen on 15.06.15.
//  Copyright © 2015 RHC. All rights reserved.
//

import Foundation
import CoreData
import OSLog

@objc(CDPlot)
open class CDPlot: NSManagedObject {


    open class func newOrExistingPlot( _ content: [String:String], forStation station:CDStation, inManagedObjectContext managedObjectContext: NSManagedObjectContext) -> CDPlot {
        if let unwrapped = content["DataID"], let dataId = Int(unwrapped) {
            let request: NSFetchRequest<CDPlot> = CDPlot.fetchRequest()
            request.predicate = NSPredicate(format: "dataId == %@ and station == %@", argumentArray: [dataId, station])
            request.fetchLimit = 1

            do {
                let result = try managedObjectContext.fetch(request)
                if let plot = result.first {
                    return plot
                }
            } catch {
            }
        }

        let entity = CDPlot.entity()
        let plot = CDPlot(entity: entity, insertInto: managedObjectContext)

        plot.updateWithContent(content)

        return plot
    }


    open class func updatePlots(_ plots: [[String: String]]) throws {
        if plots.isEmpty {
            return
        }

        guard
            let plot = plots.first,
            let stationString = plot["StationID"],
            let stationId = Int(stationString)
        else {
            return
        }

        let context = DataManager.shared.container.newBackgroundContext()

        context.mergePolicy = NSOverwriteMergePolicy

        let station = try CDStation.existingStationWithId(stationId, inManagedObjectContext: context)
        let insertedPlots = NSMutableSet()

        for plot in plots {
            let managedObject = CDPlot.newOrExistingPlot(plot, forStation: station, inManagedObjectContext: context)

            if managedObject.isInserted {
                insertedPlots.add(managedObject)
            }
        }

        if insertedPlots.count > 0 {
            station.willChangeValue(forKey: "plots")
            station.addToPlots(insertedPlots)
            station.didChangeValue(forKey: "plots")
        }

        try context.save()
    }

    func updateWithContent( _ content: [String:String] ) {
        if let unwrapped = content["Time"] {
            self.plotTime = DataManager.shared.dateFromString(unwrapped)
        }

        if let unwrapped = content["WindAvg"], let avg = Double(unwrapped) {
            self.windAvg = avg as NSNumber?
        }

        if let unwrapped = content["WindMax"], let max = Double(unwrapped) {
            self.windMax = max as NSNumber?
        }

        if let unwrapped = content["WindMin"], let min = Double(unwrapped) {
            self.windMin = min as NSNumber?
        }

        if let unwrapped = content["DirectionAvg"], let dir = Double(unwrapped) {
            self.windDir = dir as NSNumber?
        }

        if let unwrapped = content["Temperature1"], let temp = Double(unwrapped) {
            self.tempAir = temp as NSNumber?
        }

        if let unwrapped = content["DataID"], let dataId = Int(unwrapped) {
            self.dataId = dataId as NSNumber?
        }
    }


    @objc open func windDirectionString() -> String {
        let bundle = AppConfig.sharedConfiguration.frameworkBundle
        var direction = self.windDir!.floatValue

        if direction > 360.0 || direction < 0.0 {
            direction = 0.0;
        }

        switch(direction) {
        case 0...11.25, 347.75...360.0:
            return NSLocalizedString("DIRECTION_N", bundle: bundle, comment: "WindPlot, Wind direction: N")
        case 11.25...33.35:
            return NSLocalizedString("DIRECTION_NNE", bundle: bundle, comment: "WindPlot, Wind direction: NNE")
        case 33.75...56.25:
            return NSLocalizedString("DIRECTION_NE", bundle: bundle, comment: "WindPlot, Wind direction: NE")
        case 56.25...78.75:
            return NSLocalizedString("DIRECTION_ENE", bundle: bundle, comment: "WindPlot, Wind direction: ENE")
        case 78.75...101.25:
            return NSLocalizedString("DIRECTION_E", bundle: bundle, comment: "WindPlot, Wind direction: E")
        case 101.25...123.75:
            return NSLocalizedString("DIRECTION_ESE", bundle: bundle, comment: "WindPlot, Wind direction: ESE")
        case 123.75...146.25:
            return NSLocalizedString("DIRECTION_SE", bundle: bundle, comment: "WindPlot, Wind direction: SE")
        case 146.25...168.75:
            return NSLocalizedString("DIRECTION_SSE", bundle: bundle, comment: "WindPlot, Wind direction: SSE")
        case 168.75...191.25:
            return NSLocalizedString("DIRECTION_S", bundle: bundle, comment: "WindPlot, Wind direction: S")
        case 191.25...213.75:
            return NSLocalizedString("DIRECTION_SSW", bundle: bundle, comment: "WindPlot, Wind direction: SSW")
        case 213.75...236.25:
            return NSLocalizedString("DIRECTION_SW", bundle: bundle, comment: "WindPlot, Wind direction: SW")
        case 236.25...258.75:
            return NSLocalizedString("DIRECTION_WSW", bundle: bundle, comment: "WindPlot, Wind direction: WSW")
        case 258.75...281.25:
            return NSLocalizedString("DIRECTION_W", bundle: bundle, comment: "WindPlot, Wind direction: W")
        case 281.25...303.75:
            return NSLocalizedString("DIRECTION_WNW", bundle: bundle, comment: "WindPlot, Wind direction: WNW")
        case 303.75...326.25:
            return NSLocalizedString("DIRECTION_NW", bundle: bundle, comment: "WindPlot, Wind direction: NW")
        case 326.25...348.75:
            return NSLocalizedString("DIRECTION_NNW", bundle: bundle, comment: "WindPlot, Wind direction: NNW")
        default:
            return NSLocalizedString("DIRECTION_UKN", bundle: bundle, comment: "WindPlot, Wind direction: UKN")
        }
    }
}
